components/permissions/google_sign_in/google_sign_in_permission_util.cc

```diff
@@ -0,0 +1,202 @@
+#include <vector>
+
+#include "base/callback_helpers.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/functional/callback_helpers.h"
+#include "brave/components/constants/pref_names.h"
+#include "brave/components/permissions/google_sign_in/features.h"
+#include "components/permissions/permissions_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "extensions/common/url_pattern.h"
+#include "third_party/blink/public/mojom/permissions/permission_status.mojom-shared.h"
+
+namespace permissions {
+
+namespace {
+constexpr char kGoogleAuthPattern[] =
+    "https://accounts.google.com/o/oauth2/auth*";
+constexpr char kFirebaseContentSettingsPattern[] =
+    "https://[*.]firebaseapp.com/__/auth*";
+// ContentSettingsPattern accepts [*.] as wildcard for subdomains, while
+// URLPattern takes *.
+constexpr char kFirebaseUrlPattern[] = "https://*.firebaseapp.com/__/auth*";
+
+bool IsGoogleAuthUrl(const GURL& gurl, const bool check_origin_only) {
+  static const std::vector<URLPattern> auth_login_patterns({
+      URLPattern(URLPattern::SCHEME_HTTPS, kGoogleAuthPattern),
+      URLPattern(URLPattern::SCHEME_HTTPS, kFirebaseUrlPattern),
+  });
+  return std::any_of(auth_login_patterns.begin(), auth_login_patterns.end(),
+                     [&gurl, check_origin_only](URLPattern pattern) {
+                       if (check_origin_only) {
+                         return pattern.MatchesSecurityOrigin(gurl);
+                       }
+                       return pattern.MatchesURL(gurl);
+                     });
+}
+}  // namespace
+
+bool IsGoogleAuthRelatedRequest(const GURL& request_url,
+                                const GURL& request_initiator_url) {
+  return IsGoogleAuthUrl(request_url, false) &&
+         !IsGoogleAuthUrl(request_initiator_url, true) &&
+         !net::registry_controlled_domains::SameDomainOrHost(
+             request_initiator_url,
+             GURL(URLPattern(URLPattern::SCHEME_HTTPS, kGoogleAuthPattern)
+                      .host()),
+             net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+}
+
+bool IsGoogleSignInFeatureEnabled() {
+  return base::FeatureList::IsEnabled(
+      permissions::features::kBraveGoogleSignInPermission);
+}
+
+bool IsGoogleSignInPrefEnabled(PrefService* prefs) {
+  return prefs->FindPreference(kGoogleLoginControlType) &&
+         prefs->GetBoolean(kGoogleLoginControlType);
+}
+
+blink::mojom::PermissionStatus GetCurrentGoogleSignInPermissionStatus(
+    content::PermissionControllerDelegate* permission_controller,
+    content::WebContents* contents,
+    const GURL& request_initiator_url) {
+  return permission_controller->GetPermissionStatusForOrigin(
```

Comment:
I beleive you don't need to pass `request_initiator_url`. Use `GetPermissionStatusForCurrentDocument` method instead.